### ISSUE TITLE
Some security considerations

### ISSUE THREAD

**Issue by paul90** [dobbs/wiki-plugin-frame#2]
To protect wiki from unintended consequences, we really need to enable some basic practice about how content is included with `<iframe>`.

I suggest that https://github.com/dobbs/wiki-plugin-frame/blob/master/client/frame.js#L36 is modified to:

```
 $item.append('<iframe sandbox="allow-scripts"></iframe><p></p>');
```

`sandbox` is a whitelist, some of the other tokens might be useful - see [MDN <iframe>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe). **BUT**, I would not like to see `allow-same-origin` added.

An alternative approach, at the risk of making things more complex, would be to use a DSL to configure the sandbox. With this approach allow the sandbox to be configured as needed, but it would also allow `allow-scripts` and `allow-same-origin` mutually exclusivity to be enforced.

**Comment by WardCunningham** [comment]
I agree we should be cautious. As I read through the various scams and phishing attacks that arrive by email each day I ask myself, how prepared are we to be successful enough that those kinds of people pay attention to wiki?

**Comment by dobbs** [comment]
Thanks for pointing to the `sandbox` option, @paul90.  As I've been experimenting, I find I need both `allow-scripts` and `allow-same-origin` enabled in order to run apparatus in a frame.  I suspected the same would be true for something like Observable and upon testing discovered a different issue: #3

The combination of `allow-scripts` and `allow-same-origin` presents the risk of privilege escalation when both framed content an its parent live on the same origin.  I can instrument `emit` to detect when the framed content is in the same origin as the parent page.  Another option is to consider how we serve files from the assets plugin... what options to consider like `X-Frame-Options` or similar things over there expressed as Content Security Policies?

**Comment by paul90** [comment]
Content Security Policies (CSP) are interesting, but something to set over in [fedwiki/wiki-server](https://github.com/fedwiki/wiki-server). I did briefly look at it when beaker changed their browser defaults....

**Comment by dobbs** [comment]
I've added the sandbox attribute to iframes.  With this commit I've made it harder for authors/programmers like me who are inclined to hack around with javascript in assets.  I'm choosing that in exchange for reducing the chance that a non-programmer author might copy hostile javascript to their wiki via the combination of assets and frames plugins.

I think this matches the current authorities... a site owner can upload assets, but javascript in those assets can't sneak into the wiki UX via this frame plugin.

**Comment by dobbs** [comment]
Also worth saying out loud... I've closed this issue, but don't intend for that to close the discussion.  Feel free to add comments if you are so inclined.

**Comment by dobbs** [comment]
Today, @WardCunningham updated the code hosting http://scad.fed.wiki and updated plugins as well.  The upgrade broke display of his 3d designs.  The error message pictured below shows that it was this security fix which broke the 3d renderings.

![image](https://user-images.githubusercontent.com/297766/44668001-8f2eb380-a9d9-11e8-8951-f3eaaaf5fee1.png)

Re-reading the discussion here, I still find our rationale for caution compelling.  What should we recommend to authors in light of this most recent example?

In my commit message where I introduced this security measure I offered this work-around:

> Authors interested in adding new behavior to wiki are encouraged to
> create plugins rather than using frames in combination with assets.

Perhaps it is time to revisit @paul90's original suggestion:

> An alternative approach, at the risk of making things more complex, would be to use a DSL to configure the sandbox. With this approach allow the sandbox to be configured as needed, but it would also allow allow-scripts and allow-same-origin mutually exclusivity to be enforced.

We could introduce markup in the plugin to override the current behavior:
```
I TRUST THE AUTHOR OF THIS LOCAL ASSET
```

**Comment by dobbs** [comment]
The complexity in the sandbox attribute has made it hard to reason about the security of the frame plugin. This solution sets the sandbox to always be `allow-scripts` (matching @paul90's original suggestion 😄 ).

@WardCunningham, I have also tested these pages in a cross-origin scenario:
:one: http://scad.fed.wiki/tablet-clamp.html
:two: http://growing.ward.bay.wiki.org/pattern-diagrams.html

:one: works without modification.  You could safely update the frame plugin there without breaking the pages displaying your 3d models.

:two: works, provided I add the domain name to the URL specified.  Updating the frame plugin here would require updating pages.

In this change, we choose to accept the risk that scripts inside assets in the same origin, when shared through a frame, will be able to modify the DOM of the hosting wiki. I think this is really no more risky than plugmatic plugin in that plugins can only be installed by admins, and similarly assets can only be uploaded  by admins.

**Comment by dobbs** [comment]
I am also going to look at #4 before I publish the updates.  I may decide there to relax the requirement that a domain name be provided.

**Comment by dobbs** [comment]
I have fixed #4 without relaxing the domain requirement.  I will re-open #1 instead.
